node/primitives/src/lib.rs

```diff
@@ -287,6 +291,96 @@ pub struct AvailableData {
 	pub validation_data: PersistedValidationData,
 }
 
+/// This is a convenience type to allow the ErasureCode to Decode into a nested BoundedVec
+#[derive(PartialEq, Eq, Clone, Debug, Hash)]
+pub struct Proof(BoundedVec<BoundedVec<u8, 1, MERKLE_NODE_MAX_SIZE>, 1, MERKLE_PROOF_MAX_DEPTH>);
+
+impl Proof {
+	fn as_vec(&self) -> Vec<Vec<u8>> {
+		self.0.as_vec().iter().map(|v| v.as_vec().clone()).collect()
+	}
+}
+
+impl TryFrom<Vec<Vec<u8>>> for Proof {
+	type Error = &'static str;
+
+	fn try_from(input: Vec<Vec<u8>>) -> Result<Self, Self::Error> {
+		if input.len() > MERKLE_PROOF_MAX_DEPTH {
+			return Err("Merkle max proof depth exceeded.")
+		}
+		let mut out = Vec::new();
+		for element in input.into_iter() {
+			let data: BoundedVec<u8, 1, MERKLE_NODE_MAX_SIZE> =
+				BoundedVec::from_vec(element).map_err(|_| "Merkle node max size exceeded.")?;
+			out.push(data);
+		}
+		Ok(Proof(BoundedVec::from_vec(out).expect("Buffer size is deterined above. QED")))
+	}
+}
+
+impl Decode for Proof {
+	fn decode<I: Input>(value: &mut I) -> Result<Self, CodecError> {
+		match value.remaining_len()? {
+			Some(l) if l % MERKLE_NODE_MAX_SIZE != 0 =>
+				return Err("At least one intermediate node in Proof is unpadded.".into()),
+			Some(l) if (l / MERKLE_NODE_MAX_SIZE) > MERKLE_PROOF_MAX_DEPTH =>
+				return Err("Supplied proof depth exceeds maximum proof depth.".into()),
+			None => return Err("Buffer is empty.".into()),
+			_ => {},
+		}
+		let mut out = Vec::new();
+		while let Ok(Some(_)) = value.remaining_len() {
+			let mut temp = [0u8; MERKLE_NODE_MAX_SIZE];
+			value.read(&mut temp)?;
+			let bounded_temp: BoundedVec<u8, 1, MERKLE_NODE_MAX_SIZE> =
+				BoundedVec::from_vec(temp.to_vec())
+					.expect("Buffer is specifically set to bounded vector's upper bound. QED");
+			out.push(bounded_temp);
+		}
+		BoundedVec::from_vec(out).map(Self).map_err(|_| {
+			"Length of proof is ensured to be less than bouned. This Error in unreachable".into()
+		})
+	}
+}
+
+impl Encode for Proof {
+	fn size_hint(&self) -> usize {
+		MERKLE_NODE_MAX_SIZE * MERKLE_PROOF_MAX_DEPTH
+	}
+
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+		let mut out = Vec::new();
+		for element in self.0.iter() {
+			let temp = element.as_vec();
+			let mut element_vec = [0u8; MERKLE_NODE_MAX_SIZE];
```

Comment:
This is not the tersest way to store Merkle proofs, but it works. I wanted to have something working quickly in order to get some of the higher-level questions out of the way.

Strangely, the Merkle proofs generated in the polkadot tests seem to always an outer length 2, which I currently don't fully understand. The inner length varies from 3 to 347 and I also don't fully understand why.

Based on how we store these proofs as a Vec<Vec<u8>>, we can choose the best way to bound the inner and out vectors.